File: src/main/java/com/kgaft/VulkanLib/Device/DeviceBuilder.java
package com.kgaft.VulkanLib.Device;

import java.util.ArrayList;
import java.util.Collection;

public class DeviceBuilder {
    private ArrayList<String> requestExtensions = new ArrayList<>();
    private boolean requirePresentSupport = false;
    private boolean requireGraphicsSupport = false;
    private boolean requireComputeSupport = false;
    private long surfaceForPresentationCheck;
    private boolean rayTracingSupport = false;

    public void addExtension(String extensionName) {
        requestExtensions.add(extensionName);
    }
    public void addExtensions(Collection<String> extensions){
        requestExtensions.addAll(extensions);
    }
    public void requestGraphicSupport(){
        requestExtensions.add("VK_KHR_dynamic_rendering");
        requestExtensions.add("VK_KHR_depth_stencil_resolve");
        requestExtensions.add("VK_KHR_create_renderpass2");
        requestExtensions.add("VK_KHR_multiview");
        requestExtensions.add("VK_KHR_maintenance2");
        requireGraphicsSupport = true;
    }

    public void requestRayTracingSupport() {
        requestExtensions.add("VK_KHR_acceleration_structure");
        requestExtensions.add("VK_KHR_ray_tracing_pipeline");
        requestExtensions.add("VK_KHR_ray_query");
        requestExtensions.add("VK_KHR_pipeline_library");
        requestExtensions.add("VK_KHR_deferred_host_operations");
        requestExtensions.add("VK_KHR_spirv_1_4");
        requestExtensions.add("VK_EXT_descriptor_indexing");
        requestExtensions.add("VK_KHR_buffer_device_address");
        requestExtensions.add("VK_KHR_shader_float_controls");
        requestExtensions.add("VK_KHR_maintenance3");
        requestExtensions.add("VK_KHR_device_group");
        rayTracingSupport = true;
    }

    public void requestComputeSupport() {
        requireComputeSupport = true;
    }

    public void requestPresentSupport(long surface) {
        surfaceForPresentationCheck = surface;
        requirePresentSupport = true;
        requestExtensions.add("VK_KHR_swapchain");
    }

    public ArrayList<String> getRequestExtensions() {
        return requestExtensions;
    }

    public boolean isRequirePresentSupport() {
        return requirePresentSupport;
    }

    public boolean isRequireGraphicsSupport() {
        return requireGraphicsSupport;
    }

    public boolean isRequireComputeSupport() {
        return requireComputeSupport;
    }

    public long getSurfaceForPresentationCheck() {
        return surfaceForPresentationCheck;
    }

    public boolean isRayTracingSupport() {
        return rayTracingSupport;
    }
}

File: src/device/device_builder.rs
use ash::vk;
use std::ffi::{c_char, CString};
use crate::util::c_string_vec_to_ptr;

#[derive(Default)]
pub struct VlDeviceBuilder {
    request_extension: Vec<CString>,
    require_present: vk::SurfaceKHR,
    require_graphics: bool,
    require_raytracing: bool,
    require_compute: bool,
}

impl VlDeviceBuilder {

    pub fn new() -> Self {
        Self{..Default::default()}
    }
    pub fn add_extension(&mut self, extension: CString) {
        self.request_extension.push(extension);
    }

    pub fn enable_presentation(&mut self, surface: vk::SurfaceKHR) {
        self.require_present = surface;
    }

    pub fn enable_raytracing(&mut self) {
        self.request_extension
            .push(CString::new("VK_KHR_acceleration_structure").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_ray_tracing_pipeline").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_pipeline_library").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_deferred_host_operations").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_spirv_1_4").unwrap());
        self.request_extension
            .push(CString::new("VK_EXT_descriptor_indexing").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_buffer_device_address").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_shader_float_controls").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_maintenance3").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_device_group").unwrap());
        self.require_raytracing = true;
    }

    pub fn enable_graphics(&mut self) {
        self.request_extension
            .push(CString::new("VK_KHR_dynamic_rendering").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_depth_stencil_resolve").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_create_renderpass2").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_multiview").unwrap());
        self.request_extension
            .push(CString::new("VK_KHR_maintenance2").unwrap());
        self.require_graphics = true;
    }

    pub fn get_extensions(&self) -> Vec<*const c_char> {
        c_string_vec_to_ptr(&self.request_extension)
    }

    pub fn enable_compute(&mut self) {
        self.require_compute = true;
    }

    pub fn require_present(&self) -> vk::SurfaceKHR {
        self.require_present
    }

    pub fn require_graphics(&self) -> bool {
        self.require_graphics
    }

    pub fn require_raytracing(&self) -> bool {
        self.require_raytracing
    }

    pub fn require_compute(&self) -> bool {
        self.require_compute
    }

    pub fn request_extension(&self) -> &Vec<CString> {
        &self.request_extension
    }
}
